### PR TITLE
[XPU] Use a fused GDN kernel from sgl-kernel for Qwen3.5

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
## Motivation

To use a fused GDN kernel provided by sgl-kernel-xpu for better performance than what the existing triton kernels in SGLang produce.

## Modifications

- Added a subclass of `GDNAttnBackend` for XPU, which does some checks and calls the fused GDN kernel.
- Added a simple dispatch mechanism for the fused GDN kernel on XPU with existing paths unchanged.

## Accuracy Tests

Qwen3.5-9B, GSM8K, 200 samples

Fused GDN kernel from sgl-kernel-xpu
- Accuracy: 0.910
- Invalid: 0.000

Triton GDN in SGLang
- Accuracy: 0.900
- Invalid: 0.000

## Speed Tests and Profiling

Qwen3.5-9B on a Intel Arc BMG60 machine
```
python -I -m sglang.bench_one_batch --batch-size 1 --input 4096 --output 1024 --model Qwen/Qwen3.5-9B --tp 1 --device xpu --mem-fraction-static 0.9 --attention-backend intel_xpu --linear-attn-backend intel_xpu --disable-radix-cache --profile --profile-stage all --profile-activities CPU XPU --chunked-prefill-size -1 --max-prefill-tokens 20480 --max-total-tokens 20480
```

Latency | Old | New
-- | -- | --
Prefill E2E | 0.937 s | 0.808 s
Prefill GDN | 165 ms | 59.17 ms
Decode E2E | 0.0404 s | 0.0402 ms
Decode GDN | 0.522 ms | 0.452 ms

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #32383609683](https://github.com/sgl-project/sglang/actions/runs/32383609683)
Latest PR Test (Extra): :hourglass_flowing_sand: [Run #32688872423](https://github.com/sgl-project/sglang/actions/runs/32688872423)
Latest PR Test (AMD ROCm 7.2): :x: [Run #32383610036](https://github.com/sgl-project/sglang/actions/runs/32383610036)

### CODE REVIEW

File: python/sglang/srt/layers/rotary_embedding/mrope.py [hunk 1]
```diff
@@ -305,6 +305,7 @@ def forward_xpu(
         assert positions.ndim in (1, 2)
         self._match_cos_sin_cache_dtype(query)
         if positions.ndim == 2 and self.mrope_section:
+            self._match_cos_sin_cache_dtype(query)
```
Comment (Xia-Weiwen) [inline python/sglang/srt/layers/rotary_embedding/mrope.py]: This is a fix of data type mismatch. It is not related to the GDN kernels. But without this fix, models fail to run.

**Comment by Xia-Weiwen** [comment]
Hi @mingfeima  @airMeng Can you please review? Thanks.

File: python/sglang/srt/hardware_backend/xpu/attention/xpu_gdn_backend.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+from sgl_kernel import gdn_attention as sgl_kernel_gdn_attention
+
+from sglang.srt.layers.attention.linear.gdn_backend import GDNAttnBackend
+from sglang.srt.layers.radix_linear_attention import RadixLinearAttention
+from sglang.srt.model_executor.forward_batch_info import ForwardBatch
+from sglang.srt.utils import get_bool_env_var
+
+# Opt-in: dispatch GDN to the fused vLLM SYCL op in sgl-kernel-xpu
+# (torch.ops.sgl_kernel.gdn_attention). Default ON. Set to OFF to use the triton path.
+_xpu_fused_gdn_enabled = get_bool_env_var("SGLANG_XPU_FUSED_GDN", "True")
```
Comment (mingfeima) [inline python/sglang/srt/hardware_backend/xpu/attention/xpu_gdn_backend.py]: do not use `get_bool_env_var`, the current rule is register  `EnvBool` in `environ.py` and use `envs.SGLANG_XPU_FUSED_GDN.get()` for it.

**Review by mingfeima — CHANGES_REQUESTED** [review]
update UT `test_xpu_qwen3_5_9b.py` to cover both fast path and triton path:

since this pr will default enable fast path

something like:

```
class Qwen3_5_9BXPUBase(SimpleEvalGSM8KXPUMixin, CustomTestCase):
    model = "Qwen/Qwen3.5-9B"
    tp_size = 4
    accuracy = 0.90
    num_examples = 50
    num_threads = 4
    max_tokens = 8192
    other_args = SimpleEvalGSM8KXPUMixin.other_args + [
        "--page-size",
        "128",
        "--max-total-tokens",
        "65536",
        "--mem-fraction-static",
        "0.85",
    ]


class TestQwen3_5_9BXPUFusedGDN(Qwen3_5_9BXPUBase):
    pass


### test triton path here
class TestQwen3_5_9BXPUGDNFallback(Qwen3_5_9BXPUBase):
    env = {"SGLANG_XPU_FUSED_GDN": "False"}
    num_examples = 8
    accuracy = 0.0
```

File: python/sglang/srt/hardware_backend/xpu/attention/xpu_gdn_backend.py (hunk 2 above)
Comment (airMeng) [reply python/sglang/srt/hardware_backend/xpu/attention/xpu_gdn_backend.py]: should we always use the sgl-kernel GDN as default?

Comment (Xia-Weiwen) [reply python/sglang/srt/hardware_backend/xpu/attention/xpu_gdn_backend.py]: Update. Thanks.

Comment (Xia-Weiwen) [reply python/sglang/srt/hardware_backend/xpu/attention/xpu_gdn_backend.py]: > should we always use the sgl-kernel GDN as default?

The default is `true` and we keep this flag for convenience (e.g. compare SYCL and Triton kernels).

Comment (mingfeima) [reply python/sglang/srt/hardware_backend/xpu/attention/xpu_gdn_backend.py]: already default from the code i think

**Review by mingfeima — APPROVED** [review]

File: python/sglang/srt/models/qwen3_5.py [hunk 3]
```diff
@@ -636,39 +637,55 @@ def forward(
...
-                num_v_heads_tp = triton.cdiv(self.num_v_heads, self.attn_tp_size)
-            mixed_qkv, z, b, a = fused_qkvzba_split_reshape_cat_contiguous(
+        core_attn_out = z = None
+        from sglang.srt.model_executor.forward_context import get_attn_backend
+
+        backend = get_attn_backend()
+        backend = getattr(backend, "linear_attn_backend", backend)
+        if hasattr(backend, "forward_fused_gdn") and backend.supports_fused_gdn(
+            self.attn, forward_batch
+        ):
```
Comment (mingfeima) [inline python/sglang/srt/models/qwen3_5.py]: doing such a backend chosen in a `forward` seems not appropriate.

how does the original code choose the backend? when it is settled?

Comment (Xia-Weiwen) [reply python/sglang/srt/models/qwen3_5.py]: The original path does not use a fused kernel (CUDA, CPU, etc.). So, they just go to the `else` branch. We need to intercept here because we call a fused kernel.

Comment (Xia-Weiwen) [reply python/sglang/srt/models/qwen3_5.py]: Updated. Now we use a separate `_forward_xpu`.

File: python/sglang/srt/models/qwen3_5.py [hunk 4]
```diff
@@ -628,6 +629,52 @@ def _forward_input_proj_fused_quant_amd(self, hidden_states):
...
+            core_attn_out_pad = torch.zeros_like(z)
+            core_attn_out_pad[: core_attn_out.shape[0], :] = core_attn_out
+            core_attn_out = core_attn_out_pad
+
+        core_attn_out = self.norm(core_attn_out, z)
+        core_attn_out = core_attn_out.reshape(z_shape_og)
+        core_attn_out = core_attn_out.reshape(*core_attn_out.shape[:-2], -1)
+
+        output, _ = self.out_proj(core_attn_out)
+        return output
```
Comment (mingfeima) [inline python/sglang/srt/models/qwen3_5.py:695]: `forward()` calls `_forward_xpu`.

then runs `_forward_input_proj`, then often returns None (default backend is still Triton). 

`forward()` then runs _forward_input_proj again. 

Every GDN layer does input projection twice on the default path. 

you can check supports_fused_gdn (or is_intel_xpu()) before the first projection.

**Review by mingfeima — CHANGES_REQUESTED** [review]

File: python/sglang/srt/models/qwen3_5.py (hunk 4 above)
Comment (Xia-Weiwen) [reply python/sglang/srt/models/qwen3_5.py:695]: Thanks. I have moved the `_forward_xpu` to after the projection.

File: test/registered/xpu/llm_models/test_xpu_qwen3_5_9b.py [hunk 5]
```diff
@@ -37,5 +39,25 @@ class TestQwen3_5_9BXPU(SimpleEvalGSM8KXPUMixin, CustomTestCase):
...
+    dispatch, not a full accuracy regression test (already covered by the
+    default Triton path above); accuracy threshold is left at 0 accordingly.
+    """
+
+    other_args = Qwen3_5_9BXPUBase.other_args + [
+        "--linear-attn-backend",
+        "intel_xpu",
+    ]
+    num_examples = 8
+    accuracy = 0.0
```
Comment (mingfeima) [inline test/registered/xpu/llm_models/test_xpu_qwen3_5_9b.py]: using accuracy = 0.0 / 8 examples, let wrong tokens pass. keep a real GSM8K threshold on the fused class (e.g. 50 / 0.90).

File: python/sglang/srt/layers/rotary_embedding/mrope.py [hunk 6]
```diff
@@ -308,6 +308,7 @@ def forward_xpu(
...
```
Comment (mingfeima) [inline python/sglang/srt/layers/rotary_embedding/mrope.py]: `_match_cos_sin_cache_dtype` is called twice.

File: python/sglang/srt/models/qwen3_5.py [hunk 7]
```diff
@@ -643,6 +687,13 @@ def forward(
             hidden_states
         )
 
+        if _is_xpu:
+            xpu_out = self._forward_xpu(
+                projected_states_qkvz, projected_states_ba, forward_batch
+            )
+            if xpu_out is not None:
+                return xpu_out
```
Comment (mingfeima) [inline python/sglang/srt/models/qwen3_5.py]:
```suggestion
if _is_xpu and get_exec().mamba.linear_attn_backend == "intel_xpu":
    backend = get_attn_backend()
    backend = getattr(backend, "linear_attn_backend", backend)
    if backend.supports_fused_gdn(self.attn, forward_batch):
        return self._forward_xpu(
            backend, projected_states_qkvz, projected_states_ba, forward_batch
        )
```

Comment (mingfeima) [reply python/sglang/srt/models/qwen3_5.py]: ```if xpu_out is not None``` is not very straightforward, suggest make it more readable.

File: python/sglang/srt/models/qwen3_5.py [hunk 8]
```diff
@@ -628,6 +629,49 @@ def _forward_input_proj_fused_quant_amd(self, hidden_states):
...
+    ):
+        core_attn_out = z = None
+        from sglang.srt.model_executor.forward_context import get_attn_backend
+
+        backend = get_attn_backend()
+        backend = getattr(backend, "linear_attn_backend", backend)
+        if not hasattr(backend, "forward_fused_gdn") or not backend.supports_fused_gdn(
+            self.attn, forward_batch
+        ):
+            return None
```
Comment (mingfeima) [inline python/sglang/srt/models/qwen3_5.py]:
```suggestion
```

File: python/sglang/srt/models/qwen3_5.py [hunk 9]
```diff
@@ -628,6 +629,49 @@ def _forward_input_proj_fused_quant_amd(self, hidden_states):
...
         return projected_states_qkvz, projected_states_ba
 
+    def _forward_xpu(
+        self,
+        projected_states_qkvz: torch.Tensor,
+        projected_states_ba: torch.Tensor,
+        forward_batch: ForwardBatch,
+    ):
+        core_attn_out = z = None
+        from sglang.srt.model_executor.forward_context import get_attn_backend
```
Comment (mingfeima) [inline python/sglang/srt/models/qwen3_5.py]:
```suggestion
```

File: python/sglang/srt/models/qwen3_5.py [hunk 10]
```diff
@@ -628,6 +629,49 @@ def _forward_input_proj_fused_quant_amd(self, hidden_states):
             projected_states_ba, _ = self.in_proj_ba(hs_bf16)
         return projected_states_qkvz, projected_states_ba
 
+    def _forward_xpu(
+        self,
+        projected_states_qkvz: torch.Tensor,
+        projected_states_ba: torch.Tensor,
+        forward_batch: ForwardBatch,
+    ):
+        core_attn_out = z = None
```
Comment (mingfeima) [inline python/sglang/srt/models/qwen3_5.py]: can we also delete `core_attn_out = z = None`?

**Review by mingfeima — CHANGES_REQUESTED** [review]
some minor issues to address.

**Comment by mingfeima** [comment]
@Xia-Weiwen could you please check with validation team why xpu ci failed again?

File: test/registered/xpu/llm_models/test_xpu_qwen3_5_9b.py (hunk 5 above)
Comment (Xia-Weiwen) [reply test/registered/xpu/llm_models/test_xpu_qwen3_5_9b.py]: Update. Thanks.

File: python/sglang/srt/layers/rotary_embedding/mrope.py (hunk 6 above)
Comment (Xia-Weiwen) [reply python/sglang/srt/layers/rotary_embedding/mrope.py]: Update. Thanks.

File: python/sglang/srt/models/qwen3_5.py (hunk 7 above)
Comment (Xia-Weiwen) [reply python/sglang/srt/models/qwen3_5.py]: Update. Thanks.

File: python/sglang/srt/models/qwen3_5.py (hunk 8 above)
Comment (Xia-Weiwen) [reply python/sglang/srt/models/qwen3_5.py]: Update. Thanks.

File: python/sglang/srt/models/qwen3_5.py (hunk 9 above)
Comment (Xia-Weiwen) [reply python/sglang/srt/models/qwen3_5.py]: Update. Thanks.

File: python/sglang/srt/models/qwen3_5.py (hunk 10 above)
Comment (Xia-Weiwen) [reply python/sglang/srt/models/qwen3_5.py]: Update. Thanks.

**Comment by Xia-Weiwen** [comment]
> @Xia-Weiwen could you please check with validation team why xpu ci failed again?

CI should be good now.

**Review by mingfeima — APPROVED** [review]
